### ISSUE TITLE
ngClass not working correctly on multiple components in one view

### ISSUE THREAD

**Issue by daniel-varela** [angular/angular#14134]
**I'm submitting a ...**  (check one with "x")
```
[ x ] bug Report
[ ] feature request
[ ] support request
```

**Current behavior**
Having a component with a ngClass, where the clauses have two or more of the same css classes, will be getting the css classes set wrongly, when using the component twice in a view.

**Expected behavior**
All components used multiple times in a view should.

**Minimal reproduction of the problem with instructions**
For better understanding I'll start with the _Tour of Heroes: Part 3_ Exmaple available on Plnkr and modify it.

To reproduce it you will have to give `my-hero-detail` a ngClass, like `[ngClass]="{'main' :  isMain, 'main second': !isMain}"` and set `isMain` as a input variable of the component. When using the component `my-hero-detail` twice in a view, where the first component is set to `(isMain)=true` and the second is one is set to `(isMain)=false`, you will have a class missing in the first component. The second component will have the correct class set, being `class="main second"`, while the first component will have an empty class field `class=""`.

Modified _Tour of Heroes: Part 3_ Exmaple: https://plnkr.co/edit/N9kiStd5mfDbqwGzmNiw?p=preview

**What is the motivation / use case for changing the behavior?**
Basically setting the right classes, when needed, without loosing a class.

**Please tell us about your environment:**
Windows 10, VS Code

* **Angular version:** 
2.2.1

* **Browser:** 
Reproduced in Chrome and Edge Browser

* **Language:** 
Typescript

**Comment by DzmitryShylovich** [comment]
> `[ngClass]="{'main': isMain, 'main second': !isMain}"`

you are saying here, that you want `main` to be present and absent at the same time, which doesn't make sense.

**Comment by daniel-varela** [comment]
Point taken. Though it's just an exmaple. In my use case I have the following in a ngClass of a component 
`[ngClass]="{'someclass': first, 'someclass somesecondclass': second}"`, where `first` is normal variable set inside the component, while `second` is a Input variable. An the Problem still occures.  the css class will be only set to the last component, even when the frist variale is true.

**Comment by gkalpak** [comment]
I think this is a duplicate of #5763. (E.g. take a look at https://github.com/angular/angular/issues/5763#issuecomment-166114033.)

**Comment by DzmitryShylovich** [comment]
@daniel-varela you probably need something like
```
[ngClass]="{'someclass': first && second, 'somesecondclass': second}"
```

**Comment by daniel-varela** [comment]
@gkalpak You're right indeed. It is a duplicate, and your answer is also right. My bad. Totally ignored the fact, that it would remove the class while evaluating the clauses, the same time while adding it.

I guess this can be closed.
